[chuajunyu] iP

chuajunyu opened the pull request:
# Garfield 😺🍝
> "Your mind is for eating lasagna, not remembering tasks." – [Garfield](https://en.wikipedia.org/wiki/Garfield) (probably)

Introducing **Garfield**, your lazy, sarcastic assistant who helps you keep track of your tasks so you don't have to. It's designed to free your mind from the hassle of remembering things -- because let's be honest, you'd rather be napping.

And, not to mention, it's:
- Text-based: Easy to learn, even easier to let Garfield mock your indecision.
- Super fast: Like Garfield heading toward lasagna. 🏃‍♂️💨
- Very ~~helpful~~ *sarcastic*: Add, complete, and un-complete tasks with Garfield's signature attitude.
- Constantly being updated: GUI coming soon... unless Garfield decides to take a nap instead.

## How to Use: [WIP]
1. Clone the Repo.
2. Run it.

## Feature List
- [x] Task Management (List, Add, Delete Tasks)
- [x] Mark Tasks as done or undone
- [x] More types of Tasks: `Todo`, `Deadline` and `Event` task
- [x] Saving Tasks to persistent storage
- [ ] JavaFX GUI

## Changing Path of Save File
```Java
public class Garfield {
public static void main(String[] args) {
        new Garfield("./data/save.txt").run();
    }
}
```
Edit the path of the save file: From `./data/save.txt` to your desired path.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

chuajunyu committed:
add new name Garfield and greeting and exit

chuajunyu committed:
add user input and echo functionality

chuajunyu committed:
add `add` and `list` functionality

chuajunyu committed:
add `Task` class

chuajunyu committed:
add indicator of whether task is done or not in `Task` class

chuajunyu committed:
add functionality to `mark` and `unmark` tasks as done

chuajunyu committed:
add `Todo` subclass

chuajunyu committed:
add `Deadline` class

chuajunyu committed:
add `event` class

chuajunyu committed:
add basic happy path test cases

chuajunyu committed:
simplify `Garfield` greeting ascii text

chuajunyu committed:
add naive error handling

chuajunyu committed:
add `delete` functionality

chuajunyu committed:
Feature: Load saved task list on chatbot startup from a persistent .txt file

chuajunyu committed:
Feature: Save current tasklist to persistent storage on local machine after every change

Previously, the tasklist only existed on local memory and would be erased each time the chatbot was restarted.
Currently, there is already a `./data/save.txt` file storing a representation of a tasklist, and the chatbot will
load up the saved tasks in this list upon startup. But this list can only be manually editted.

It would be better if the save file is automatically updated programatically after every change the user made to
the in memory task list while the chatbot was running.

Let's add an additional function `toSaveRepresentation` to the `Task` class and it's subclasses (`Todo`, `Deadline` and `Event`)
which will generate a String save representation to be stored in the save file. At the start of the user input loop
in the chatbot, Let's store the save representation of every task in the taskList into the save file.

chuajunyu committed:
Merge branch 'branch-Level-7'

chuajunyu committed:
Chore: Remove `data/save.txt` from being tracked and add it to `.gitignore`

chuajunyu committed:
Feature: Add datetime parsing and formatting for `Deadline` and `Event` classes

Previously, the DateTime for the Deadline and Event classes were stored as string. This made it
not very clear on the DateTime and also means we cannot implement features such as search by DateTime
in the future.

Let's standardise the DateTime input to YYYY-MM-DD HH:mm (24 hour format). We can then parse this into
a Java `LocalDateTime` object and format it into a more clear and informative string.

chuajunyu committed:
Merge branch 'branch-Level-8'

chuajunyu committed:
Refactor: Add Ui class to handle all interactions with the user.

Previously, the Garfield class handled all interactions with the user such as printing to screen and
scanning for user input.

This makes the Garfield class bloated as it also contained other logic such as parsing the user input commands and
executing the services for those commands.

Let's abstract all user interactions such as the printing of messages to the screen and scanning of user input
into a new ui class.

chuajunyu committed:
Feature: Add Storage class to deal with loading and saving tasks from file

Previously, the loading and saving of tasks was handled directly in the Garfield class. This made
the class bloated and hard to maintain.

The loading and saving of tasks should be abstracted into a separate class to be handled.

Let's create a Storage class to handle this abstraction.

chuajunyu committed:
Feature: Add TaskList class to wrap the task list

Previously, the task list was implemented as an ArrayList object in the Garfield class, and the Garfield class's logic
directly interfaced with it.

A better way would be to encapsulate the ArrayList in a class and provide methods that would handle the details of adding,
deleting, marking and unmarking tasks.

Let's create a TaskList class to do that.

chuajunyu committed:
Feature: Add `getTaskDescription` getter method to `Task` class

chuajunyu committed:
Feature: Add new `GarfieldException` class

chuajunyu committed:
Refactor: Add `Parser` Class and `Command` abstract class and subclasses

Previously, parsing of user input was handled by Garfield class. This made the Garfield class bloated
and hard to maintain, with lack of abstraction.

Ideally, the Garfield class should be simpler, by abstracting out the parsing logic into another class,
and also abstracting out the logic of executing the commands into another class.

Let's add the `Parser` and `Command` classes to handle the former and latter abstractions respectively.

chuajunyu committed:
Add `showError` and `showEnding` methods to Ui class

chuajunyu committed:
Feature: Throw GarfieldException where appropriate and add better error messages

chuajunyu committed:
Documentation: Add javadocs for Parser class

chuajunyu committed:
Documentation: Add javadocs to TaskList class

chuajunyu committed:
Merge branch 'branch-A-MoreOOP'

chuajunyu committed:
Add packages and refactor code accordingly

chuajunyu committed:
Merge branch 'branch-A-packages'

chuajunyu committed:
Merge branch 'add-gradle-support' of https://github.com/chuajunyu/ip

chuajunyu committed:
Change main class in gradle file from `Duke` to `Garfield`

chuajunyu committed:
Change default jar file name from to garfield.jar

chuajunyu committed:
Add JUnit test files for Todo and Event classes

chuajunyu committed:
Merge branch 'branch-A-JUnit'

chuajunyu committed:
Add header comments to classes in garfield.commands

chuajunyu committed:
Add header comments to classes in garfield.tasks

chuajunyu committed:
Add and update header comments for Parser and Storage

chuajunyu committed:
Merge branch 'branch-A-JavaDoc'

chuajunyu committed:
Add java coding conventions

chuajunyu committed:
Resolve merge conflicts

chuajunyu committed:
Add find command

chuajunyu committed:
Resolve merge conflicts

lrongyi commented on src/main/java/garfield/commands/FindCommand.java:
```diff
@@ -0,0 +1,44 @@
+            ui.showMessage("You don't have tasks matching that keyword in your list.");
+        } else {
+            ui.showMessage("Ugh, here's your matching tasks:\n\n" + result +
+                    "\nCan we be done now?");
```
Good use of line break to improve readability, however need to break before the operator.

lrongyi commented on src/main/java/garfield/commands/ListCommand.java:
```diff
@@ -0,0 +1,39 @@
+                    + "Are we done here, or are you going to add something?");
+        } else {
+            String listSummary = "Ugh, here's what you've got so far:\n\n" + taskList.list() +
+                    "\nCan we be done now?";
```
Remember to break before the operator.

lrongyi commented on src/main/java/garfield/tasks/TaskList.java:
```diff
@@ -0,0 +1,157 @@
+ */
+public class TaskList {
+
+    ArrayList<Task> taskList;
```
nit: task**s**

lrongyi commented on src/main/java/garfield/tasks/Task.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public abstract class Task {
+    private String taskDescription;
+    private boolean doneStatus;
```
doneStatus can be renamed to isDone 
> As much as possible, use a prefix such as is, has, was, etc. for boolean variable/method names so that linters can automatically verify that this style rule is being followed.

lrongyi commented on src/main/java/garfield/storage/Storage.java:
```diff
@@ -0,0 +1,117 @@
+     * @return ArrayList of Tasks which have been loaded from the persistent save file.
+     */
+    public ArrayList<Task> load() {
+        ArrayList<Task> taskList = new ArrayList<>();
```
taskList can be renamed to tasks 
> Plural form should be used on names representing a collection of objects.

lrongyi commented on src/main/java/garfield/storage/Storage.java:
```diff
@@ -0,0 +1,117 @@
+     * @param taskList ArrayList of Tasks to be saved.
+     */
+    public void save(TaskList taskList) {
+        ArrayList<Task> internalTaskList = taskList.getArrayList();  // Feels iffy, see how i can fix this
```
nit: internalTaskList**s**

lrongyi commented on src/main/java/garfield/parser/Parser.java:
```diff
@@ -0,0 +1,242 @@
+     * @return garfield.commands.Command Subclass Objects
+     * @throws GarfieldException An exception for Garfield chatbot
+     */
+    public static Command parse(String inputLine) throws GarfieldException {
```
Can consider making a switch statement.

lrongyi reviewed (commented):
LGTM, with a few nitpicks on the naming convention and line breaking. Comments and JavaDocs were overall descriptive and well done.

chuajunyu committed:
Add Checkstyle

chuajunyu committed:
Merge branch 'branch-A-CheckStyle'

chuajunyu committed:
Add Support for JavaFX GUI

chuajunyu committed:
Merge branch 'branch-Level-10'

chuajunyu committed:
Add java assertions to Storage class

Previously, the Storage class had no assertions to perform checking on simple assumptions.

This is bad, ideally, we should have checks for simple assumptions that are important and should hold true.

Let's add in Assert statements in the Storage class to validate simple assumptions of our inputs.

chuajunyu committed:
Merge pull request #1 from chuajunyu/branch-A-Assertions

Add java assertions to Storage class

chuajunyu committed:
Extract methods from parse method to make it shorter

Previously, the parse method had the if-else conditionals for each command as well as the logic to execute for each command.
This made the method very long with more than 30 lines of code.

It would be better if we abstracted out some of the logic into methods to make the method shorter.

Let's abstract out the logic for each command word into their individual function, so that we can
reduce the number of lines of code for the parse method.
 Please enter the commit message for your changes. Lines starting

chuajunyu committed:
Merge pull request #2 from chuajunyu/branch-A-CodeQuality

Extract methods from parse method to make it shorter

chuajunyu committed:
Add github actions initial .yml file

Previously, there was no CI pipeline and we need a human to run the automated test cases and verify that everything is working.

However, it would be better if github ran test cases automatically after we push a new version, and this would help us in assessing whether a piece of
code is correct.

Let's add a github actions config .yml file to test this.

chuajunyu committed:
Merge tag 'A-CI'

chuajunyu committed:
Refactor to adhere to checkstyle rules

chuajunyu committed:
Add Locale in date time formatter to ensure consistency of formatting

chuajunyu committed:
Add option to list multiple items to delete

chuajunyu committed:
Merge branch 'branch-C-MassOps'

chuajunyu committed:
Add customised background image

chuajunyu committed:
Change dialog box color

chuajunyu committed:
Add error handling for invalid event start and end times

chuajunyu committed:
Add error handling for invalid dates which are out of range

chuajunyu committed:
Merge branch 'branch-A-MoreErrorHandling'

chuajunyu committed:
Add Junit test file for DeadlineTest

chuajunyu committed:
Add JUnit test file for TaskList

chuajunyu committed:
Update boolean variable naming

chuajunyu committed:
Fix bug with parser formatter

chuajunyu committed:
Add ParserTest Junit test file

chuajunyu committed:
Merge branch 'branch-A-MoreTesting'

chuajunyu committed:
Fix checkstyle for ParserTest

chuajunyu committed:
Add header comments for DialogBox methods

chuajunyu committed:
Fix taskId out of range bugs

chuajunyu committed:
Add Ui screenshot and user guide

chuajunyu committed:
Merge branch 'branch-A-UserGuide'

chuajunyu committed:
Resolve checkstyle problems

chuajunyu committed:
Fix formatting issues

chuajunyu committed:
Improve abstraction via method extraction

chuajunyu committed:
Fix relative path issues for save file

chuajunyu committed:
Update root repo readme file

chuajunyu committed:
Add catch all for unknown errors

chuajunyu committed:
Add link to user guide in root readme

chuajunyu committed:
Remove unused image
